sonarts-core/src/runner/sonartsServer.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * SonarTS
+ * Copyright (C) 2017-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as net from "net";
+import * as path from "path";
+import * as os from "os";
+import * as fs from "fs";
+import * as ts from "typescript";
+import { getIssues } from "./rules";
+import { FileCache, createService } from "./languageService";
+import { parseTsConfig } from "../utils/parser";
+
+export function start(port: number) {
+  const EMPTY_ANSWER = '{"issues":[]}';
+  const fileCache = new FileCache();
+  // key is ts file path, value is corresponding tsconfig path
+  const tsConfigCache: Map<string, string> = new Map();
+  const servicesPerTsconfig: Map<string, ts.LanguageService> = new Map();
+
+  const server = net.createServer(socket => {
+    socket.on("data", data => {
+      const request = JSON.parse(data.toString());
+
+      if (request.operation === "analyze") {
+        const { file, rules, content } = request;
+
+        fileCache.newContent({ file, content });
+
+        let tsConfig;
+        if (tsConfigCache.has(file)) {
+          tsConfig = tsConfigCache.get(file)!;
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-351&issues=AWD5U5dREFfjxkQcPdcY&open=AWD5U5dREFfjxkQcPdcY)

[ISSUE_KEY]: AWD5U5dREFfjxkQcPdcY

Reply:
This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED